doc/development/powershell.md

```diff
@@ -0,0 +1,165 @@
+
+- Reserve powershell pipelines (`|` operator) for simple assignment operations only (e.g. light transformations of a data structure)
+  - If side effects are needed, favor `foreach` loops.
+  - If the value being returned is any sort of collection (e.g. list), either favor a `foreach` loop or explicitly declare a variable type like `[array]$arr = some-func | ? { $_ -contains "foo" }` . Due to how pipelines work, when the result value is singular, the inner value with the collection item type will be returned instead of a collection with length 1. _This goes the same for powershell functions._
+- Handle exit codes from calls to external binaries (e.g. `git`).
+  - Powershell by default will only stop execution on thrown exceptions. Calls to external binaries do not throw powershell native exceptions, so failures will not be caught unless explicitly handled.
+    - Check the `$LASTEXITCODE` variable for values greater than 0.
+    - Use the new `$PSNativeCommandUseErrorActionPreference` variable coming in 7.3 (https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-73?view=powershell-7.2#experimental-features).
+- Set global powershell variables to achieve consistent behavior:
+  - `$ErrorActionPreference` should be set to `Stop` at the top of any cmdlet files unless there is a specific reason to do something else.
+  - `Set-StrictMode -Version 4` should be set at the top of any cmdlet files unless there is a specific reason to do something else.
+- Support the common `-WhatIf` parameter. See [SupportsShouldProcess docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#supportsshouldprocess).
+  - Scripts should be able to be run locally, and they should also support a dry run mode via `-WhatIf` so a user can see what actions might be performed that could be destructive.
+  - Most common powershell module functions will support `-WhatIf` natively, but engsys code that does any write operations should also gate this functionality behind this parameter by checking [ShouldProcess](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#pscmdletshouldprocess).
+- Log write actions with enough context that someone reading the log statement in a pipeline would be able to know what happened.
+  - For example, `Write-Host "Updating code owner to '$user' in '$ServiceDirectory'"` is a much better log than `Write-Host "Updating codeowner"`.
+
+## Style
+
+Powershell is a very flexible language. In engsys we try to use a smaller subset of what's possible to keep programs consistent and their behavior predictable.
+
+- Declare types for function parameters (e.g. `function foo([string]$bar) {}`.
+- Avoid complex pipeline statements.
+- Put curly braces `{` for opening function blocks on a new line.
```

Comment:
IMO, this is not good. ScriptBlocks must start on the same line as the cmdlet unless you use line continuation, which looks funny. Named ScriptBlocks (i.e. functions) can have them start on a new line, but now you have a hodgepodge of styles:

```powershell
function foo($str)
{
  Write-Host $str
}
1..10 | foreach-object {
  foo $_
}
```
Just keep it consistent: use JavaScript-style bracing and start every open brace on the same line.

Reply:
This one is a @weshaggard preference for a more C# style, but I think at the time of discussion we agreed that moving all blocks onto new lines would be too much given the existing code.

Reply:
You also can't for cmdlets. Try typing this:

```pwsh
1..10 | foreach-object
{
```
After hitting Enter on the first line, you'll be prompted for the `Process` argument. That's my point: when calling cmdlets, you need to put the brace on the same line. Functions can go either way, so for a consistent script I recommend always using JavaScript-style bracing i.e. start on the same line.

Reply:
There are always expectations to the rules (single line expressions which is what you essentially highlighting) but in general for larger blocks of code I think having it on its own line makes things more readable. However for smaller blocks my care level is less.

Reply:
If you put the `{` on the expression line it'll work 100% of the time unlike putting `{` on the subsequent line. Especially for devs who don't know PowerShell as well, why go through the pain? Personally, I don't see this making the code less readable. PowerShell shares a lot of syntax in common with C#, sure, but it's not C#. If anything, many of the idioms were based on Python (this has been mentioned in many older documents since the days of "Monad").

Reply:
I do agree with @heaths here. I don't think the braces on a newline is consistent with general usage of the language or with other similar conventions in a way that's intuitive to someone reading or writing these scripts.

Reply:
I'm Ok with not having this in the guidelines. I do like consistency and with the expressions I agree we cannot have 100% consistency. However for functions or larger blocks I still would be OK with having it on its own line (because I still like it for readability) but I'm Ok without it as well.